tensorboard/data/server/run.rs

```diff
@@ -171,9 +188,33 @@ impl RunLoader {
     ///
     /// If we need to access `run_data` but the lock is poisoned.
     pub fn reload(&mut self, filenames: Vec<PathBuf>, run_data: &RwLock<commit::RunData>) {
+        let run_name = self.run.0.clone();
+        debug!("Starting load for run {:?}", run_name);
+        let start = Instant::now();
         self.update_file_set(filenames);
-        self.reload_files();
-        self.commit_all(run_data);
+        let mut n = 0;
+        let mut last_commit_time = Instant::now();
+        self.reload_files(|run_loader_data, event| {
```

Comment:
Thoughts on inlining this callback into `reload_files`? I don’t mind
callbacks in general, but they are a bit more complicated than not using
them. If `self.run` moves to `self.data.run`, then you only need to pass
in `self.reload_files(run_data, &start)`, which feels pretty reasonable
to me: it’s natural that `reload_files` should take the destination, and
`start` is just for instrumentation.

Then `reload` reads like:

```rust
let start = Instant::now();

self.update_file_set(filenames);
self.reload_files(run_data, &run_name, &start);
self.data.commit_all(run_data);
// ^ maybe move this into `reload_files`, too? maybe; could go either way

debug!(
    "Finished load for run {:?} ({:?})",
    self.data.run.0,
    start.elapsed()
);
```

Reply:
Hmm.  So I arrived at this because I liked `reload_files()` better with the business logic of what to do with the event and how often to commit abstracted out of it.  In particular part of my reasoning was that if we wanted to parameterize the 5 second interval or do something fancier than a fixed interval (like triggering based on the event details itself), that configuration would all need to be plumbed into `reload_files()` either as their own parameters or as fields on `self`, and encapsulating it via the callback felt prettier.

As a halfway solution, I originally did have `commit_all()` called in `reload_files()` and was trying to pass in some kind of smaller "check if I need to commit yet" predicate instead, but then I remembered that abstractions are supposed to be free in Rust so I figured hey maybe I can just get away with calling a callback closure on every event :)

I'll grant it makes `reload()` a little messier; but I liked being able to read the sequence of commit steps and the timing details all in one function.  So if you aren't particularly opposed I think I'm happier with it this way.

Reply:
This also seems reasonable to me. Sold.

> abstractions are supposed to be free in Rust so I figured hey maybe I can just get away with calling a callback closure on every event :)

[@chandlerc might contort that there are no zero-cost abstractions][1],
and that (Rust) closures have no run-time cost but do have human cost.
:-) But yeah, I’m okay with them here.

[1]: https://youtu.be/rHIkrotSwcc?t=9

Reply:
No such thing as a free lunch after all huh 😿